osf/models/collection.py

```diff
@@ -0,0 +1,153 @@
+
+class CollectedGuidMetadata(BaseModel):
+    primary_identifier_name = 'guid___id'
+
+    class Meta:
+        order_with_respect_to = 'collection'
+        unique_together = ('collection', 'guid')
+
+    collection = models.ForeignKey('Collection', on_delete=models.CASCADE)
+    guid = models.ForeignKey('Guid', on_delete=models.CASCADE)
+    creator = models.ForeignKey('OSFUser')
+    collected_type = models.CharField(blank=True, max_length=31)
+    status = models.CharField(blank=True, max_length=31)
+
+    @cached_property
+    def _id(self):
+        return self.guid._id
+
+class Collection(GuidMixin, BaseModel):
+    objects = IncludeManager()
+
+    provider = models.ForeignKey('AbstractProvider', blank=True, null=True, on_delete=models.CASCADE)
+    creator = models.ForeignKey('OSFUser')
+    guid_links = models.ManyToManyField('Guid', through=CollectedGuidMetadata, related_name='collections')
+    collected_types = models.ManyToManyField(
+        'contenttypes.ContentType',
+        related_name='+',
+        limit_choices_to={
+            'model__in': ['abstractnode', 'basefilenode', 'collection', 'preprintservice']
+        })
+    title = models.CharField(max_length=200, validators=[validate_title])
+    collected_type_choices = ArrayField(models.CharField(max_length=31), blank=True, default=list)
+    status_choices = ArrayField(models.CharField(max_length=31), blank=True, default=list)
+    is_public = models.BooleanField(default=False, db_index=True)
+    is_promoted = models.BooleanField(default=False, db_index=True)
+    is_bookmark_collection = models.BooleanField(default=False, db_index=True)
+    is_deleted = models.BooleanField(default=False, db_index=True)
+
+    def __unicode__(self):
+        return '{self.title!r}, with guid {self._id!r}'.format(self=self)
+
+    @property
+    def url(self):
+        return '/{}/'.format(self._id)
+
+    def get_absolute_url(self):
+        return self.absolute_api_v2_url
+
+    @property
+    def absolute_api_v2_url(self):
+        return api_v2_url('/collections{}'.format(self.url))
+
+    @property
+    def linked_nodes_self_url(self):
+        return self.absolute_api_v2_url + 'relationships/linked_nodes/'
+
+    @property
+    def linked_registrations_self_url(self):
+        return self.absolute_api_v2_url + 'relationships/linked_registrations/'
+
+    @property
+    def linked_nodes_related_url(self):
+        return self.absolute_api_v2_url + 'linked_nodes/'
+
+    @property
+    def linked_registrations_related_url(self):
+        return self.absolute_api_v2_url + 'linked_registrations/'
+
+    def save(self, *args, **kwargs):
+        first_save = not self.id
+        if self.is_bookmark_collection:
+            if self.title != 'Bookmarks':
+                # Bookmark collections are always named 'Bookmarks'
+                self.title = 'Bookmarks'
+            if not self.pk and self.creator.collection_set.filter(is_bookmark_collection=True, is_deleted=False).exists():
```

Comment:
minor, but should this go before title checking (fail faster)

Reply:
No queries would be avoided by doing this earlier. But on principle, sure